README.md
# Spooky-Tweaks

Spooky Tweaks is a large scale vanilla overhaul mod that serves as an 'unofficial update' to the game. That is to say, it rebalances, fixes, streamlines, implements unused content, and reduces annoyance. While it attempts to be compatible with content mods, it may conflict with other vanilla-altering mods.

This mod adds no __new__ save-altering content! You can add and remove it at any time without destroying your saves, but remember to unstack items with increased stack sizes first. Please also remember to always make save backups when using mods.

This project is open source. You may create a modular fork if you only want a certain feature or small set of features. It would be appreciated if these are called 'Spooky Modules', but you have the freedom to do what you'd prefer.

### Addons
- [Spooky Integrations](https://github.com/LeDrascol/Spooky-Integrations)
- Spooky Additions (Unreleased)

### Project Links
- [Steam Workshop](https://steamcommunity.com/sharedfiles/filedetails/?id=737895230)
- [Chucklefish Forums](https://community.playstarbound.com/resources/spooky-tweaks.3931/)
- [Starbound Nexus](https://community.playstarbound.com/resources/spooky-tweaks.3931/) (Not Maintained)
- [Logo Artist](https://steamcommunity.com/id/EvKem)

# Project Highlights
## General
- Moon related things are now spooky
- Mr Spooks (Erchius Ghost) moves faster, can be knocked back, and dies in lava
- Main missions 'clue' hunting can be skipped by scanning the gate
- Teleporting has no artificial time padding
- More fuel types, higher fuel values, and higher fuel costs for FTL travel
- The sewing machine makes SIGNIFIGANTLY (90%) less noise
- Vehicles are one-handed, and their costs rebalanced
- Grappling hooks are accessible earlier and use physics just like ropes
- Reward bags stack
- Bandages and antidotes can be crafted at any time
- Late-storyline outpost shops will indicate if they open later

**Updates**
- No more startup animation
- Play modes are now NORMAL, EXPERT, and HARDCORE
- 66 new facial emotion triggers for chat
- Portable Pixel Printer is back and functional
- Radio messages don't provide constant arbitrary praise
- More optional Outpost clues to find
- SAIL tells you which EPP you need for dangerous planets
- Shockhopper can drop a trophy
- EPPs use more scientific names
- Hoverbikes hover much better and usually won't cause fall damage
- Status Pods now have nine new possible effects
- Mazebound64 moved to Hylotl arcades
- Quest text updated to remove arbitrary praise
- Immense overhaul of rarity levels to reflect item value
- Amazing new features added to the Starbound Arcade Machine (removed before releasing)
- Five new blessings from Tonauac (outpost bird)
- Lava immunity status effect is now obtainable as Magma Mitigation
- Electric aura status effect is now obtainable as Energy Aura
- All instruments are now slightly cheaper and psychic (playing multiple at once will overload your brain)
- New racial descriptions as part of Spooky Tweaks' gradual usurping of the lore
- Teleporters are now platforms that can be walked past
- Items with no interaction or placement ability moved to materials tab
- Glow effects reworked for better ambient lighting
- Learn any race's flag by picking it up
- Flashlight beams are almost identical to the scanner's light

**Latest Update**
- Racial bonuses for Avian, Human, Familiar, Glitch, and Hylotl
- Splash screen is back, but silent and won't interrupt loading
- Human codexes follow the mod's lore tweaks
- Uncensored names for breakable alcohol bottles
- Tents block hunger just like beds
- New tooltip information to show mining speeds for tools
- Jump height increased to between that of release and beta
- Teleportation only takes 0.2 seconds, with no fake loading times
- No fake loading time for the Bounty Board

## Economy
- New Shop: Chip Tunes - Selling all vanilla musical instruments (replaces Mazebound64)
- Teleporter cores are bought with money instead of diamonds
- Buy and sell rates are more reasonable (average 100% to buy, 80% to sell)
- Upgrade parts can now be purchased at the Tech Dispenser (Outpost)
- Capture pods can be purchased at The Outpost
- Penguin Weapon Shop sells rare and otherwise unobtainable weapons

**Updates**
- Treasured Trophies sells more rare items such as:
-- Fishaxe, Challenge Doors, Weapon Upgrade Anvil
-- Doomcanon, Terraforge
- Infinity Express sells new novelty items
- Weapon shop now sells boss items after beating the respective boss
- Canned food costs 21x more due to the chocolate shortage (and immense value)
- Minor outpost shops have more unique names
- Prize counters now sell plushies and toys for tickets
- Prize tickets are given out as rewards from loot bags
- Station vending machines reworked to sell mech salvage
- Buy your own personal Rob Repairo
- ~Mech blueprints can be traded for at Beakeasy~
- Mech Assembly Station can be purchased from the tech dispenser
- Terramart Shipments buys items for their full value (Even with Enhanced Storage)
- Terramart Shipments accept all mining materials
- Lana's supply business has improved to compete with augment crafting
- Vending machines don't buy things from you
- Peacekeeper shop items moved to places that make more sense
- Fake ship licenses are 50% more expensive
- Ore values are approximately 4x higher
- Ursa Miner sells radioactive rods usable as ship fuel
- Teleporters require cores AND relevant materials
- Five new otherwise unobtainable sets in Frogg Furnishings
- 49 more unused items sold at the outpost

**Latest Update**
- Beakeasy is an actual drinks serving bar
- Hyltol prize counter joins the outpost, with 70+ new categorized items
- Terramart sells the Relocator
- Penguin weapon shop sells Neo weapons
- Ancient items move to the Terraforge instead of Treasured Trophies
- New safer method to remove shop items
- Overhaul to Chip Tunes with tabs and mod support

## Tech
- Tech descriptions are more helpful
- Sonic sphere is faster and easier to control
- Blink dash recharges as fast as the normal dash
- Rocket jumping is almost instant, launches further, and recharges mid-air
- Multi-Jump goes slightly higher
- Sprinting is slightly faster

**Updates**
- Sprinting won't restrict using other tech or items
- All distortion spheres have enhanced handling and negate fall damage
- Aqua Sphere can move freely in water instead of on top as a boat
- Blink dash can blink further, in mid-air, and without landing
- Wall jumping refreshes it's double jump ability
- Spike Sphere seriously for real this time won't fall off things
- Tech missions require tech cards instead of ingots

**Latest Update**
- Fixes and cleanup for sphere tech

## Balance
- All staves and wands charge significantly faster
- Healing zones from staves heal much faster
- Mission rewards gradually increase in magnitude
- Erchius Facility always gives 20 Erchius Crystals (1000 Total Fuel)
- Challenge doors stay around twice as long (Reduced in Update 2)
- Teleporters are not destroyed when broken
- Tech consoles are not destroyed when broken
- Money loss removed for casual
- Value, speed, and power of unique weapons raised to be more worthwhile
- Creatures can be attacked while frozen in time
- Racial storage lockers hold 64 items instead of 48

**Updates**
- Rocket Spear ability rebalanced to into a mobility tool
- Final mission reward bag provides incredible loot
- Protector's Broadsword buffed to compete with unique weapons
- Dual wielding chainsaws for extreme lumbering
- EPPs emit a small amount of light
- Backpacks that should provide air actually do (ex. Oxygen Tank)
- Wings provide jump boosts equal to an augment
- AA Battery Pack adds additional energy capacity
- Toxic Flower Backpack will block poison effects
- Toxic Waste Backpack is more dangerous then ever
- Medic Backpack gives a small regeneration effect
- Mech battery pickups heal twice as much energy
- High tech mech parts balanced out to create aesthetic choice
- Solarium tier mech arms buffed to match endgame weapon strength
- Shield drone can now deploy up to five orbiting shields
- All mech drones last longer (2-5 minutes) and have reduced cooldowns
- Simple mech parts improved to be worth using
- Outpost mech missions now actually give rewards
- Shawl of Elation is now more uplifting (low gravity)
- Avian god wings now hold a mysterious ancient power
- Foods that claimed to be spooky are now actually spooky
- Ship lights are harder to accidentally break
- Drill Spear mobility buffed for maximum heavens piercing potential
- Mech arm attachments reworked for usability
- Higher tier mech bodies are more energy efficient
- Ship lockers make food rot 80% slower

**Latest Update**
- No dropping items on death in Survival mode, but money drop restored to 30%
- Mining tools don't use durability, and drills mine in a 3x3 area
- Adaptable crossbow requires less energy to fire
- Broken Broadsword attacks faster and provides a defensive ability

## Crafting
- Augments can be crafted at the Research Station, and specify their exact functions
- Restored crafting for disabled items such as Perfect Armor, Tall Chair, and Valentines Heart Forge
- 3D printing for ancient themed items and vending machines

**Updates**
- Energy pickaxe is craftable and slightly improved
- Mech Part Crafting Table is craftable like any workstation
- Player station transponders now cost the same as any other device
- Player stations require exponentially less resources to upgrade
- Restored crafting for Toymaker's Table and glow grenades
- More Protectorate items at the Industrial Workbench
- Boss doors are now printable for all your one-way needs
- The orange stimpack returns
- Mining tools can be made at an Agricultural Station
- 62 more unused items made craftable
- Stimpacks moved from the Medical Station to the Apothecary

**Latest Update**
- Water coolers dispense water into empty bottles
- Pickaxes must be crafted, rather than found as loot or purchased
- All Campfire crafting is accessible from the Kitchen Counter
- Components are craftable in the tier two furnace, including RAM
- Many improvements to learn on pickup
- Soda is craftable like other foods

# Official Integration
The mods listed below add optional new APIs used by this mod
- [Many Frogs](https://steamcommunity.com/sharedfiles/filedetails/?id=1111078158)
- [The Tabula Rasa](https://steamcommunity.com/sharedfiles/filedetails/?id=737353165)

The mods listed below include special compatibility checks
- [Legendary Tooltips](https://steamcommunity.com/sharedfiles/filedetails/?id=740863523)


Changes.md
# Initial Release
+ Price change on Teleporter Core from 15 DIAMONDS to 2000 PIXELS
+ Changed crafting for all 2-Stop Teleshop teleporters to relevant ingredients
+ Crafting for Penguin Suit moved from Treasured Trophies to Beakeasy
+ Price change on Dubloon from 3 DIAMONDS to 5000 PIXELS
+ Added Agricultural Table crafting for Damage Augment
+ Added Agricultural Table crafting for Energy Augment
+ Added Agricultural Table crafting for Fire Block Augment
+ Added Agricultural Table crafting for Gravity Augment
+ Added Agricultural Table crafting for Healing Augment
+ Added Agricultural Table crafting for Health Augment
+ Added Agricultural Table crafting for Jump Augment
+ Added Agricultural Table crafting for Light Augment
+ Added Agricultural Table crafting for Mobily Augment
+ Added Agricultural Table crafting for Poison Augment
+ Added Agricultural Table crafting for Speed Augment
+ Added Agricultural Table crafting for Swim Augment
+ Restored Replicator crafting for Perfect Armor
+ Renamed Liquid Erchius Fuel to Liquid Spookium
+ Renamed Crystal Erchius Fuel to Crystal Spookium
+ Renamed Erchius Ghost to Mr. Spooks
+ Removed lava immunity from Erchius Ghost
+ Updated helpfulness of description on Damage Augment
+ Updated helpfulness of description on Electric Resist Augment
+ Updated helpfulness of description on Energy Augment
+ Updated helpfulness of description on Fire Resist Augment
+ Updated helpfulness of description on Healing Augment
+ Updated helpfulness of description on Health Augment
+ Updated helpfulness of description on Ice Resist Augment
+ Updated helpfulness of description on Thorns Augment
+ Price change from 12500 PIXELS to 11000 PIXELS for Red Hoverbike Controller
+ Price change from 15000 PIXELS to 11000 PIXELS for Green Hoverbike Controller
+ Price change from 25000 PIXELS to 15000 PIXELS for Green Boat Controller
+ Changed Grappling Hook projectile type to rope (gravity, longer shots, thicker)
+ Raised projectile speed from 50 to 65 for Grappling Hook
+ Changed hand-use to 1H for Torch
+ Changed hand-use to 1H for Red Hoverbike Controller
+ Changed hand-use to 1H for Green Hoverbike Controller
+ Changed hand-use to 1H for Khaki Hoverbike Controller
+ Changed hand-use to 1H for Green Boat Controller
+ Raised speed from 10 to 18 for Sonic Sphere
+ Reduced cooldown from 2 SECONDS to 0.8 SECONDS for Blink Dash
+ Updated helpfulness of description on Rocket Jump
+ Updated helpfulness of description on Wall Jump
+ Removed water jumping restrictions for Aqua Sphere
+ Updated descriptions for moons to reflect name changes on Erchius Fuels
+ Updated helpfulness of description on ancient gateway locations
+ Updated helpfulness of description on Sprint Tech
+ Allowed mid-air use on Sprint Tech
+ Raised jump force from 0.6 to 0.85 for Multi Jump
+ Raised speed from 10 to 14 for Spike Sphere
+ Reduced energy cost from 50 to 30 on Distortion Sphere
+ Reduced charge time from 0.5 to 0.05 on Rocket Jump
+ Restored fuel value of 30 FUEL to Solarium Ore
+ Restored fuel value of 75 FUEL to Solarium Star
+ Raised fuel value from 1 FUEL to 2 FUEL on Liquid Erchius Fuel
+ Raised fuel value from 2 FUEL to 6 FUEL on Crystal Erchius Fuel
+ Added fuel value of 1 FUEL to OIL
+ Fixed missing REGENERATION 1 effect on FORCE CELL (bed)
+ Reduced charge time from 1.4 to 0.6 for generated staves
+ Reduced charge time from 1.0 to 0.6 for generated wands
+ Fixed charging animation sync issues for generated staves
+ Fixed charging animation sync issues for generated wands
+ Restored Engineer's Table crafting for Heartforge
+ Price change on Perfectly Generic Object to 25 PIXELS
+ Price change on Upgrade Module to 20000 PIXELS
+ Price change on Manipulator Module from 1000 PIXELS to 2000 PIXELS
+ Price change on Tech Card from 1250 PIXELS to 1000 PIXELS
+ Price change on Penguin Costume from 10 DUBLOONS to 3 DUBLOONS
+ Crafting for Giant Bow (Head) moved from Treasured Trophies to Spinning Wheel
+ Allowed 3D printing for Nanostove
+ Crafting for Jump Pad added to Industrial Workbench
+ Crafting for Gun Chest added to Industrial Workbench
+ Removed Durasteel Bar reward from Hunting Caverns
+ Removed Iron Bar reward from Erchius Facility, Pagoda Library, Sovereign Temple, and Miniknog Stronghold
+ Raised loot entry from 1 to 3 MANIPULATOR MODULE in Erchius Facility and Hunting Caverns
+ Raised loot entry from 1 to 4 MANIPULATOR MODULE in Pagoda Library and Sovereign Temple
+ Raised loot entry from 1 to 5 MANIPULATOR MODULE in Miniknog Stronghold
+ Raised loot entry from 1 to 9 MANIPULATOR MODULE in Baron's Keep
+ Raised loot entry from 1 to 3 TECH CARD in Erchius Facility and Hunting Caverns
+ Raised loot entry from 1 to 4 TECH CARD in Pagoda Library and Sovereign Temple
+ Raised loot entry from 1 to 5 TECH CARD in Miniknog Stronghold
+ Raised loot entry from 1 to 9 TECH CARD in Baron's Keep
+ Raised loot entry from 1 to 3 DIAMOND in Erchius Facility
+ Raised loot entry from 1 to 5 DIAMOND in Hunting Caverns
+ Raised loot entry from 1 to 7 DIAMOND in Pagoda Library
+ Raised loot entry from 1 to 9 DIAMOND in Sovereign Temple
+ Raised loot entry from 2 to 3 CAPTURE POD in Erchius Facility
+ Raised loot entry from 2 to 5 CAPTURE POD in Hunting Caverns
+ Raised loot entry from 2 to 7 CAPTURE POD in Pagoda Library
+ Raised loot entry from 2 to 9 CAPTURE POD in Sovereign Temple
+ Raised loot entry from 1000 to 2500 PIXELS in Erchius Facility
+ Raised loot entry from 1000 to 3000 PIXELS in Hunting Caverns
+ Raised loot entry from 1000 to 3500 PIXELS in Pagoda Library
+ Raised loot entry from 1000 to 4000 PIXELS in Sovereign Temple
+ Updated Erchius Facility loot to always give 20 Erchius Crystals (1000 fuel)
+ Mediated all Baron's Keep loot odds to 0.1 (from 0.02/0.2/0.05/0.03)
+ Price change on Erchius Crystal to 200 PIXELS
+ Added fuel value of 50 FUEL to Erchius Crystal
+ Reduced rarity rank from Legendary to Uncommon for Reward Bag
+ Updated accuracy of description for Reward Bag

tech/jump/rocketjump_freedom.lua
require "/scripts/vec2.lua"

function init()
  self.chargeTime = config.getParameter("chargeTime")
  self.boostTime = config.getParameter("boostTime")
  self.boostSpeed = config.getParameter("boostSpeed")
  self.boostForce = config.getParameter("boostForce")
  
  -- COOLDOWN VARS
  self.rocketCooldown = config.getParameter("rocketCooldown")
  --self.rocketCooldown = 0.8 -- Value entirely too fun to use
  self.rocketCooldownTimer = 0
  
  self.rechargeEffectTimer = 0
  self.rechargeEffectTime = config.getParameter("rechargeEffectTime", 0.1)
  self.rechargeDirectives = config.getParameter("rechargeDirectives", "?fade=CCCCFFFF=0.25")
  -- COOLDOWN VARS

  idle()

  self.available = true
end

function uninit()
  idle()
end

function update(args)

  -- COOLDOWN FUNCTIONS
  if self.rocketCooldownTimer > 0 then
    self.rocketCooldownTimer = math.max(0, self.rocketCooldownTimer - args.dt)
    if self.rocketCooldownTimer == 0 then
	  -- animate
      self.rechargeEffectTimer = self.rechargeEffectTime
	  tech.setParentDirectives(self.rechargeDirectives)
      --animator.playSound("recharge")
	  animator.playSound("dashCooldown")
    end
  end
  
  if self.rechargeEffectTimer > 0 then
    self.rechargeEffectTimer = math.max(0, self.rechargeEffectTimer - args.dt)
    if self.rechargeEffectTimer == 0 then
	  -- reset
	  status.clearPersistentEffects("movementAbility")
	  self.available = true
      tech.setParentDirectives()
    end
  end
  -- COOLDOWN FUNCTIONS
  
  local jumpActivated = args.moves["jump"] and not self.lastJump
  self.lastJump = args.moves["jump"]

  self.stateTimer = math.max(0, self.stateTimer - args.dt)

  if mcontroller.groundMovement() or mcontroller.liquidMovement() then
    if self.state ~= "idle" then
      idle()
    end

    self.available = true
  end

  if self.state == "idle" then
    if jumpActivated and canRocketJump() then
      charge()
    end
  elseif self.state == "charge" then
    if self.stateTimer > 0 then
      mcontroller.controlApproachVelocity({0, 0}, self.boostForce)
      mcontroller.controlModifiers({movementSuppressed = true})
    else
      local direction = {0, 0}
      if args.moves["right"] then direction[1] = direction[1] + 1 end
      if args.moves["left"] then direction[1] = direction[1] - 1 end
      if args.moves["up"] then direction[2] = direction[2] + 1 end
      if args.moves["down"] then direction[2] = direction[2] - 1 end

      if vec2.eq(direction, {0, 0}) then direction = {0, 1} end

      boost(direction)
    end
  elseif self.state == "boost" then
    if self.stateTimer > 0 then
      mcontroller.controlApproachVelocity(self.boostVelocity, self.boostForce)
    else
      idle()
    end
  end

  animator.setFlipped(mcontroller.facingDirection() < 0)
end

function canRocketJump()
  return self.available
      and not mcontroller.jumping()
      and not mcontroller.canJump()
      and not mcontroller.liquidMovement()
      and not status.statPositive("activeMovementAbilities")
end

function charge()
  self.state = "charge"
  self.stateTimer = self.chargeTime
  self.available = false
  status.setPersistentEffects("movementAbility", {{stat = "activeMovementAbilities", amount = 1}})
  tech.setParentState("fly")
  animator.setParticleEmitterActive("rocketParticles", true)
  animator.playSound("charge")
  animator.playSound("chargeLoop", -1)
  -- COOLDOWN TIMER
  self.rocketCooldownTimer = self.rocketCooldown
  -- COOLDOWN TIMER
end

function boost(direction)
  self.state = "boost"
  self.stateTimer = self.boostTime
  self.boostVelocity = vec2.mul(vec2.norm(direction), self.boostSpeed)
  tech.setParentState()
  animator.stopAllSounds("charge")
  animator.stopAllSounds("chargeLoop")
  animator.playSound("boost")
end

function idle()
  self.state = "idle"
  self.stateTimer = 0
  status.clearPersistentEffects("movementAbility")
  tech.setParentState()
  animator.setParticleEmitterActive("rocketParticles", false)
  animator.stopAllSounds("charge")
  animator.stopAllSounds("chargeLoop")
  -- COOLDOWN AVOIDER
  self.rechargeEffectTimer = 0
  -- COOLDOWN AVOIDER
end
